### ISSUE TITLE
querier: Rate over deduplicated counter from many replicas can lead to double reset account.

### ISSUE THREAD

**Issue by bwplotka** [thanos-io/thanos#2401]
Found by GitLab, we were investigating offline with @SuperQ 

Their issue: https://gitlab.com/gitlab-com/gl-infra/infrastructure/-/issues/9293

This can be only reproducible with large rates `[30m+]` which means it has to do with chunks ordering or overlaps.

**Comment by bwplotka** [comment]
Some potential issue is that overlaps are not handled really well: https://github.com/thanos-io/thanos/pull/2400/files

**Comment by bwplotka** [comment]
It can be something related to fact that GitLab is using Store GW in HA without loadbalancer (querying both in same time), so the data is duplicated and unsorted (chunks) for sure.

**Comment by SuperQ** [comment]
@bwplotka We initially reproduced the issue with direct to Prometheus sidecars, not Store GW. These are in HA pairs as well.

**Comment by bwplotka** [comment]
Got some test data from @SuperQ so hopefully will be able to repro locally :hugs: 

Fingers crossed :crossed_fingers:

**Comment by bwplotka** [comment]
Looks like it is dup of https://github.com/thanos-io/thanos/issues/1326 - let's continue discussion there.

**Comment by bwplotka** [comment]
Actually let's not be so  sure, this might be different (here deduplication does not cause it)

**Comment by bwplotka** [comment]
@SuperQ this repro is so amazing can explore all details. Definitely we change overlapping and unsorted chunks. We should be able to find a problem in our algorithm soon, thanks!

BTW... I kind of overengineered (as you can imagine) and wrote `thanos tools storeapi serve --json=<file x>` which can serve JSON (protobuf based) and as Store API :tada: 

So I can get your file (actually anything generated by `grpcurl` and put into `thanos tools storeapi serve --json` , run querier and connect `storeapi serve` as --store, and see your results: 

![image](https://user-images.githubusercontent.com/6950331/80226032-c5578700-8643-11ea-923c-c921dd20f0d1.png)

**Comment by bwplotka** [comment]
Tooling looks like it works, but I think we don't have enough chunks to repro it :thinking: 

Tried all sorts of time ranges, steps and rate ranges.. no luck:

![image](https://user-images.githubusercontent.com/6950331/80238264-08225a80-8656-11ea-9ffe-3000e5296eb8.png)

All good everywhere...

cc @SuperQ

**Comment by SuperQ** [comment]
Here's another data sample that reproduces it. The time range to reproduce is this:

* End time: 2020-04-24 04:00
* Range: 4h
* Step: 1800
* Query: `rate(gitlab_transaction_cache_read_hit_count_total[30m])`

[data-web-08.json.gz](https://github.com/thanos-io/thanos/files/4532969/data-web-08.json.gz)
![image](https://user-images.githubusercontent.com/1320667/80280450-5db54080-8704-11ea-8e4f-230e2a90b39b.png)

**Comment by bwplotka** [comment]
I don't see full data (you gave only half of it I think, but I can repro :tada: 

![image](https://user-images.githubusercontent.com/6950331/80362824-359c1d80-887b-11ea-9491-3efbdbd572b6.png)

**Comment by bwplotka** [comment]
Investigating

**Comment by bwplotka** [comment]
Step does not matter, it's deduplication bug:

![image](https://user-images.githubusercontent.com/6950331/80363045-8ca1f280-887b-11ea-96e7-26463b4bd217.png)

**Comment by bwplotka** [comment]
Finally, found cause https://github.com/thanos-io/thanos/pull/2528

Essentially there are 3, exactly the same overlapping chunks with stale markers. For some reasons counter iterator does not handle this well.

**Comment by bwplotka** [comment]
CounterIterator must be smoking in such cases:

![image](https://user-images.githubusercontent.com/6950331/80409264-7407fb80-88c0-11ea-9ed8-0ac84110589e.png)

**Comment by bwplotka** [comment]
Reason is that most of the chunks are ignored in those cases in counter iterator.

**Comment by bwplotka** [comment]
Found root-cause, described minimal repro case: https://github.com/thanos-io/thanos/pull/2528#issuecomment-620659141

Closing https://github.com/thanos-io/thanos/issues/1326 as duplicate, now we know for sure.

**Comment by bwplotka** [comment]
Not super clear how to fix the issue long term (: Some deep dive https://docs.google.com/spreadsheets/d/13A8ChunqbVdRq9j5kqrtfzknO6mvVFQPUkXwUiBuV_4/edit?usp=sharing

**Comment by bwplotka** [comment]
Current idea is to actually have special deduplication for counters. The main issue is that we don't know what metric is a counter of on the offline level (unless it's downsampled data, then we know). In query part we can also tell as we can assume counter when rate is used.  :/

**Comment by bwplotka** [comment]
A quick win is to not worry about offline dedup yet and just implement special dedup for counters (: Then we can maybe for the offline base on `_total`.. but that's sketchy..

**Comment by bwplotka** [comment]
### TL;DR: The problem is with deduplicating a counter series from 2 or more Prometheus replicas. 

Let's say they scrape the same counter from the same application. 

Accounting resets correctly in *generic deduplication* algorithm data is really hard [as presented in this spreadsheet](https://docs.google.com/spreadsheets/d/13A8ChunqbVdRq9j5kqrtfzknO6mvVFQPUkXwUiBuV_4/edit?usp=sharing). This is due to a different view of END value for each counter by different replicas (different scrape time!).

Crafting a deduplication algorithm when we know it's counter metric is quite trivial. The problem is...  we don't know. So ideally we need generic dedup algorithm for replicas.

Any ideas @brancz @brian-brazil @beorn7 @SuperQ ? (: 

### My current idea to move this forward:

The current idea is to actually have special deduplication for counters. Generally, we don't know what metric is a counter of on the offline level (unless it's downsampled data, then we know). However, for query part it's clear. It is counter if `rate` func was in hints from PromQL. So we can use special counter-based dedup.

On offline rewrite / deduplication level, for raw data we have no idea what type is. However for a quick win, we for now, not worry about offline dedup yet and just solve query issue.

Then we can maybe for the offline figure something else. Maybe generic dedup that will work for those, or something that will base  on `_total` metric name  (but that's sketchy)

**Comment by bwplotka** [comment]
Or... we should collaborate on different dedup algorithm for future. Maybe scrape interval based? (downside: What if scrape interval changes)

**Comment by bwplotka** [comment]
Fix: https://github.com/thanos-io/thanos/pull/2548
Tests: https://github.com/thanos-io/thanos/pull/2528

**Comment by bwplotka** [comment]
Help wanted for revie!

**Comment by brian-brazil** [comment]
> (downside: What if scrape interval changes)

That's not common, but you could depend on noone having a scrape interval over 2 minutes as that's not sane for other reasons.

**Comment by bwplotka** [comment]
That's quite good idea :+1:

**Comment by beorn7** [comment]
It actually saddens me that Prometheus “by design” doesn't really cope with scrape intervals >2m. I'd love to see future Prometheus versions lifting that arbitrary limit, and I'd therefore prefer if Thanos didn't bake in that limit into its own design, too.

Interestingly, I'd also love to see future Prometheus version to have 1st class support for metric types. That would then also solve your problem of how to safely recognize a counter.
